### PR TITLE
Feautre/implement token info endpoints

### LINKED ISSUE #335: [TARA ERC-20 INFO] Implement required endpoints for new token details website
We need 3 additional routes, all of these reflecting the amounts on the Taraxa Mainnet(ETH Mainnet):

- staking
- token(already there)
- validators

**Staking endpoints**

Returns:

- Total stake (everything in the staking contract) 
- Total delegation (we have this number on the community site)
- Average validator commission % (average commission rate weighted by their delegation)

**Token endpoints**

Returns:

- Total locked (however much is still in the deployer address: 0x723304D1357A2334FCf902Aa3d232f5139080A1b)
- Total circulation (can be calculated: total supply - total locked - total staked) BUT CoinGecko needs an API so we need to give them one 

**Validators endpoints**

Returns:

- Number of block-producing mainnet validators
- Number of block-producing testnet validators 
- Total cumulative commission earned by mainnet validators (hope the community site has this)
- Total cumulative rewards earned by testnet validators (hope the community site has this)
- Total # github commits in the last 30-days (unsure if this is available without an expensive Github upgrade… I’m hoping we can all repos private + public in the taraxa project organization, right now we’re hand-counting them once a month for our update, be nice to get an automated #) https://stackoverflow.com/questions/27931139/how-to-use-github-v3-api-to-get-commit-count-for-a-repo

### PR DESCRIPTION
**Pull request by VargaElod23**
Depends on #2 .

### CODE REVIEW

File: src/staking/delegation.service.ts [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+  private readonly logger = new Logger(DelegationService.name);
+  constructor(
+    private configService: ConfigService,
+    private readonly httpService: HttpService,
+  ) {}
+
+  async totalDelegated() {
+    const delegationApi = `${this.configService.get<string>(
+      'delegationAPIRoot',
+    )}/validators?show_fully_delegated=true&show_my_validators=false`;
```
Comment (rattrap) [inline src/staking/delegation.service.ts]: Why only fully delegated validators?

File: src/staking/delegation.service.ts [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+            }),
+          ),
+      );
+      delegationData = realTimeDelegationData;
+    } catch (error) {
+      this.logger.error(error);
+      throw new InternalServerErrorException(
+        'Fetching details unsuccessful. Please try again later.',
+      );
+    }
```
Comment (rattrap) [inline src/staking/delegation.service.ts]: This whole thing can be moved to a separate method to keep things DRY

File: src/token/token.service.ts [hunk 3]
```diff
@@ -30,9 +41,70 @@ export class TokenService {
...
+    const deployerAddress = this.configService.get<string>('deployerAddress');
+    if (deployerAddress) {
+      const decimals = await this.tokenContract.decimals();
+      const totalLocked = await this.tokenContract.balanceOf(deployerAddress);
+      return totalLocked.div(BigNumber.from(10).pow(decimals)).toString();
+    } else return '0';
+  }
+  async totalCirculation() {
+    return (
+      (await this.totalSupply()) -
```
Comment (rattrap) [inline src/token/token.service.ts]: These are all strings

File: src/token/token.service.ts [hunk 4]
```diff
@@ -30,9 +41,70 @@ export class TokenService {
...
+      (await this.totalLocked()) -
+      (await this.totalStaked())
+    );
+  }
+  async stakingRatio() {
+    return (
+      ((await this.totalStaked()) /
```
Comment (rattrap) [inline src/token/token.service.ts]: Again. These methods return strings

**Review by rattrap — CHANGES_REQUESTED** [review]

File: src/staking/delegation.service.ts (hunk 1 above)
Comment (VargaElod23) [reply src/staking/delegation.service.ts]: Isn't that flag meaning: `include fully delegated validators too` ?

File: src/staking/delegation.service.ts [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    });
+    return {
+      totalDelegated: totalDelegationAcc,
+    };
+  }
+
+  async averageWeightedCommission() {
+    const delegationData = await this.fetchDelegationData();
+    let totalDelegationAcc = 0;
+    delegationData.forEach((node: any) => {
```
Comment (Malak67) [inline src/staking/delegation.service.ts:37]: We don't know the type for Node? If not then we should add node?.totalDelegation (or create a type or interfance for it).

File: src/staking/delegation.service.ts [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+      if (node.totalDelegation) {
+        totalDelegationAcc += node.totalDelegation;
+      }
+    });
+    let totalWeightedCommission = BigNumber.from(0);
+    delegationData.forEach((node: any) => {
+      if (node.totalDelegation && node.currentCommission) {
```
Comment (Malak67) [inline src/staking/delegation.service.ts:44]: Same here

File: src/staking/delegation.service.ts [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+    };
+  }
+
+  private async fetchDelegationData() {
+    const delegationApi = `${this.configService.get<string>(
+      'delegationAPIRoot',
+    )}/validators?show_fully_delegated=true&show_my_validators=false`;
+    let delegationData;
+    try {
+      const realTimeDelegationData = await firstValueFrom(
```
Comment (Malak67) [inline src/staking/delegation.service.ts]: Same comment as in the other PR.

File: src/staking/delegation.service.ts [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+          .pipe(
+            map((res) => {
+              return res.data;
+            }),
+          ),
+      );
+      delegationData = realTimeDelegationData;
+    } catch (error) {
+      this.logger.error(error);
+      throw new InternalServerErrorException(
```
Comment (Malak67) [inline src/staking/delegation.service.ts]: Remove this (see node PR)

File: src/token/token.service.ts [hunk 9]
```diff
@@ -30,9 +41,70 @@ export class TokenService {
     const decimals = await this.tokenContract.decimals();
     return decimals.toString();
   }
+  async getPrice() {
+    const taraDetailsCG = this.configService.get<string>('coinGeckoTaraxaApi');
+    let priceDetails;
+    try {
+      const realTimePriceData = await firstValueFrom(
```
Comment (Malak67) [inline src/token/token.service.ts:48]: Same here

File: src/token/token.service.ts [hunk 10]
```diff
@@ -30,9 +41,70 @@ export class TokenService {
...
+          .pipe(
+            map((res) => {
+              return res.data;
+            }),
+          ),
+      );
+      priceDetails = realTimePriceData;
+    } catch (error) {
+      this.logger.error(error);
+      throw new InternalServerErrorException(
```
Comment (Malak67) [inline src/token/token.service.ts:66]: Remove this

**Review by Malak67 — CHANGES_REQUESTED** [review]
Small changes needed

**Pull request merged**